File: Doc-Scanner-X/Doc-Scanner-X/KeyChainManager.m
//
//  KeyChainManager.m
//  Doc-Scanner-X
//
//  Created by dynamsoft on 2018/1/12.
//  Copyright © 2018年 com.dynamsoft. All rights reserved.
//

#import "KeyChainManager.h"

@implementation KeyChainManager

+ (void) saveUUID:(NSString*)uuid {
    NSDictionary* keyChainItem = [[NSDictionary alloc] initWithObjectsAndKeys:(NSString*)kSecClass,kSecAttrAccount,(NSString*)kSecValueData,uuid, nil];
    SecItemAdd((CFDictionaryRef)keyChainItem, nil);
}

+ (NSString*) readUUID {
    id result;
    NSDictionary* query = [[NSDictionary alloc] initWithObjectsAndKeys:(NSString*)kSecClass,kSecAttrAccount,(NSString*)kSecReturnData,kCFBooleanTrue,(NSString*)kSecMatchLimit, kSecMatchLimitOne, nil];
    
    SecItemCopyMatching((CFDictionaryRef)query, (CFTypeRef)&result);
    NSString* uuid = (NSString*)result;
    if (uuid == nil) {
        uuid = [UIDevice currentDevice].identifierForVendor.UUIDString;
        [self saveUUID:uuid];
        return uuid;
    } else {
        return uuid;
    }
}
@end

File: DocScannerX/DocScannerX/KeyChainManager.swift
//
//  KeyChainManager.swift
//  Doc-Scanner-X
//
//  Created by dynamsoft on 2017/11/30.
//  Copyright © 2017年 Dynamsoft. All rights reserved.
//

import Foundation
import Security


class KeyChainManager {
    
    // MARK: - Class Methods
    
    static func saveUUID(_ uuid: String){
        let keyChainItem = [kSecClass as String: kSecAttrAccount, kSecValueData as String: uuid] as [String : Any]
        SecItemAdd(keyChainItem as CFDictionary, nil)
    }
    
    /**
     Read UUID from KeyChain. If there isn't UUID in KeyChain, generate one and save it.
     - Returns: User unique ID
    */
    static func readUUID() -> String? {
        var result: AnyObject?
        
        /// Initialize the search query.
        let query = [kSecClass as String: kSecAttrAccount, kSecReturnData as String: kCFBooleanTrue, kSecMatchLimit as String: kSecMatchLimitOne] as [String: Any]

        SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer(&result))
        var uuid = result as? String
        if uuid == nil {
            uuid = UIDevice.current.identifierForVendor?.uuidString
            saveUUID(uuid!)
            return uuid
        } else {
            return uuid
        }
    }
}
